daemon: Fix detection of BPF/XDP NodePort, BPF masq and host-fw devices

brb opened the pull request:
First, this PR extends the BPF NodePort device auto-detection mechanism by:

1. In addition to a device with a default route, it considers a device with k8s Node IP addr. The k8s Node IP addr is derived from the self Node object, and it's either `InternalIP` or `ExternalIP` or `nil` (the former is preferred).
2. BPF NodePort direct routing device can be automatically detected (previously, it was the first device among all BPF NodePort devices). If there is only one BPF NodePort device is set, then that device is used for the direct routing. Otherwise, the k8s Node IP device is used. Users can specify the device via `--direct-routing-device` (`global.nodePort.directRoutingDevice`).
3. Enables XDP on the direct routing device (see reasoning in the commit msg).

The detected devices by 1. are used for BPF masquerading and host-fw too.

Second, this PR introduces `--devices` flag and deprecated `--device`. If the latter is specified, its value is appended to the former.

Reviewable per commit.

One limitation which hasn't been resolved in this PR, is that if in the k8s dualstack mode InternalIPv4 and InternalIPv6 are assigned to two different devices, then only one will be considered. I'm going to address it in a separate PR (not a release blocker).

Once this PR is merged, I will move `initKubeProxyReplacement()` to `daemon/cmd/kube_proxy.go` (I didn't want to move in this PR to avoid complicating the reviewing).

Fix: https://github.com/cilium/cilium/issues/11789

brb committed:
k8s: Store k8s Node IP addr

This commit stores k8s Node IP addr (InternalIP > ExternalIP > nil)
which is going to be used by the BPF NodePort device auto-detection.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Move kubeProxyReplacement init after connect to k8s

This is required because the NodePort BPF device detection will need
to know self k8s Node IP addrs.

Also, move BPF-masq and host-fw checks too. The both depend on settings
configured by initKubeProxyReplacementOptions().

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
vendor: Update vishvananda/netlink

Pull the latest version which includes a new field LinkIndex in Addr
struct [1]. This is going to be used for retrieving link by ip addr.

[1]: https://github.com/vishvananda/netlink/pull/543

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Move device detection into separate function

We are going to extend it, which would make
initKubeProxyReplacementOptions() less readable otherwise.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Extend BPF NodePort device auto-detection

In additional to a device with a default route, consider devices with
k8s node IP addr.

This should cover a case, when a host has two interfaces - one for
in-cluster communication, and one for outside.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm/daemon: Add --direct-routing-device

This commit adds a new agent flag "--direct-routing-device". The flag
is used in BPF NodePort in the direct routing mode.

If the flag is not set, then its value is being automatically detected.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Add unit tests for device detection

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm/daemon: Add --devices for attaching bpf_host.o

The param specifies to which devices bpf_host.o should be attached.
Currently, it's used by BPF NodePort, host-fw and BPF masquerading.

Also, mark --device as deprecated. If a user specifies both,
cilium-agent will log a fatal msg.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Drop nodeport_must_drop_remote constraint

Previously, the constraint made it impossible to have BPF NodePort
exposed via multiple devices, as a request to a NodePort service which
was received by non-XDP device (handled by TC) was dropped due to
the constraint.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Use direct routing device for XDP NodePort

Attach XDP NodePort to a device which is used for direct routing among
nodes.

In the case of a multi-dev setup, if we attach to other than the direct
routing device, a request to a NodePort svc which real destination is
a remote backend won't be redirected to another device (only possible
via hairpinning). So, we need to attach to a device which can forward
to another node which is the direct routing device.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on pkg/node/types/node.go:
```diff
@@ -223,6 +223,26 @@ func (n *Node) getNodeIP(ipv6 bool) (net.IP, addressing.AddressType) {
+		return externalIP
+	}
+
+	return nil
```
can be `return externalIP`

brb replied:
I wanted to make it more visible that `nil` can be returned as well.

joestringer replied:
Doesn't the function comment already describe this? And the declaration of the stack implies potential lack of initialization? (I don't overly care)

aanm commented on pkg/option/config.go:
```diff
@@ -2146,7 +2154,24 @@ func (c *DaemonConfig) Populate() {
+				c.Devices = append(c.Devices, device)
+			}
+		}
+	}
```
Please write this in a function

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -253,9 +253,17 @@ func init() {
-	flags.StringSliceP(option.Device, "d", []string{}, "List of devices facing cluster/external network for attaching bpf_netdev (first device should be one used for direct routing if tunneling is disabled)")
+	flags.StringP(option.Device, "d", "",
+		fmt.Sprintf("Device facing cluster/external network for attaching bpf_host. Deprecated in favor of --%s", option.Devices))
+	flags.MarkHidden(option.Device)
```
also `flags.MarkDeprecated(option.Device, "This option is no longer supported and will be removed in v1.10")`

brb replied:
It's supported, but it will be removed. Anyway, will add an description.

christarazi commented on daemon/cmd/kubeproxy_replacement_test.go:
```diff
@@ -0,0 +1,165 @@
+	s.prevConfigEnableIPv6 = option.Config.EnableIPv6
+	s.prevK8sNodeIP = node.GetK8sNodeIP()
+
+	s.currentNetNS, err = netns.Get()
```
These operations likely need to be protected by locking the goroutine to the thread via `runtime.LockOSThread`. See [netns's docs](https://pkg.go.dev/github.com/vishvananda/netns?tab=doc) for more information. See fc7716622f5d1b11753e5815472fac15ee359d61 for an example.

Relevant snippet:

> Package netns allows ultra-simple network namespace handling. NsHandles can be retrieved and set. Note that the current namespace is thread local so actions that set and reset namespaces should use LockOSThread to make sure the namespace doesn't change due to a goroutine switch. It is best to close NsHandles when you are done with them. This can be accomplished via a `defer ns.Close()` on the handle. Changing namespaces requires elevated privileges, so in most cases this code needs to be run as root.

christarazi replied:
Ah, noticed that you already do this in the code. I suggest moving the locking to the setup before each test using `SetUpTest` and unlocking in `TearDownTest`. I recognize that there's only one test right now, but if someone later expands on this suite, there's a risk they may forget, so having the locking/unlocking in the setups/teardowns would mitigate this.

brb replied:
Moving the locking to `SetUpTest` is IMHO error prone, as it's up to a testing framework to decide whether to run each test case in the same goroutine, and this might be not visible to a developer. If it end up running the test case in a different goroutine, then subtle bugs might be introduced, see https://www.weave.works/blog/linux-namespaces-golang-followup on this topic.

christarazi reviewed (changes requested):
Looks good to me, I reviewed code I'm familiar with, so didn't get to all of it. Minor comment below

joestringer commented on pkg/node/address.go:
```diff
@@ -455,3 +451,13 @@ func SetIPsecKeyIdentity(id uint8) {
+
+// SetK8sNodeIP sets k8s Node IP addr.
+func SetK8sNodeIP(ip net.IP) {
+	k8sNodeIP = ip
```
Do we need some kind of synchronization here?

joestringer replied:
(Could/Should it be able to change over time?)

brb replied:
For now noup, as getting the val happens after setting it.

joestringer commented on pkg/k8s/init.go:
```diff
@@ -196,6 +196,10 @@ func GetNodeSpec() error {
 		}
+		if option.Config.KubeProxyReplacement != option.KubeProxyReplacementDisabled {
+			log.Info("K8s node name is empty. BPF NodePort might not be able to auto detect " +
+				"devices (if not specified)")
```
Can we check `devices` to see if it's specified to clarify this info message further?

joestringer commented on pkg/k8s/init.go:
```diff
@@ -228,6 +232,9 @@ func GetNodeSpec() error {
 		node.SetLabels(n.Labels)
+
+		// K8s Node IP is used by BPF NodePort devices auto-detection
+		node.SetK8sNodeIP(n.GetK8sNodeIP())
```
What's the implication if we don't hit this case and instead fall through to the else below? 

I feel like the managed clouds were more likely to hit the `}else{` case below, but I'm not particularly familiar with this code.

brb replied:
Good concern. The `else` case will be hit, if it cannot retrieve (Cilium)Node object during the first request when both `--k8s-require-ipv{4,6}-pod-cidr` are disabled. I think it is possible only on a very busy clusters in which kube-apiserver is overloaded.

If it happens, then only a device with a default route will be able to be selected for BPF NodePort (behavior of the auto-detect in v1.6 and v1.7).

Anyway, I'm going to change `retrieveNodeInformation()` in a way that it keeps waiting for the node object if the auto-detection is required. Mainly, to avoid situations when bpf_host.o is attached to a k8s Node device only on some nodes in a busy cluster.

brb replied:
Resolved in 8bf878063d6224f9382d5859e3e5bdb7872fb4a0.

joestringer commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -370,8 +370,14 @@ data:
+{{- end }}
 {{- if .Values.global.nodePort.device }}
-  device: {{ join " " .Values.global.nodePort.device | quote }}
+  device: {{ .Values.global.nodePort.device | quote }}
```
Should this just use `devices` since `--device` is deprecated?

brb replied:
I don't have any strong opinion on this, but we currently merge both vals in `populateDevices()` in the agent for the backward compatibility. WDYT?

pchaigno replied:
Shouldn't we just reject using both `devices` and `device` (in cilium-agent or in Helm)? It feels like currently we're introducing a new behavior for an option that will then be deprecated.

brb replied:
> Shouldn't we just reject using both devices and device (in cilium-agent or in Helm)?

Yeah, makes sense. I will change it: keeping both `global.nodePort.device` and `global.devices`, but rejecting in agent if both are set.

brb replied:
Resolved in: e361f6121a31140ba2fa9e77abcab6b852bf7014.

joestringer commented on daemon/cmd/daemon_test.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-// +build !privileged_tests
+// +build linux,privileged_tests
```
IMO this is unfortunate, would be nice to mock out `bpftool` for this kind of testing if that's the constraining aspect here. I recognize that might not be a small task though.

brb replied:
Totally agree. Actually found a way how to avoid marking the suite as privileged and without mocking bpftool: https://github.com/cilium/cilium/pull/11894/commits/d8571facd67a3735eb1117b52e9c173a96a03053.

joestringer reviewed (commented):
First pass comments, I mainly looked at the logic outside of `daemon/` so far.

christarazi commented on daemon/cmd/kubeproxy_replacement_test.go:
```diff
@@ -0,0 +1,165 @@
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy1")
+}
+
+func createDummy(iface, ipAddr string) error {
+	dummy := &netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: iface,
+		},
+	}
+	if err := netlink.LinkAdd(dummy); err != nil {
+		return err
+	}
+
+	link, err := netlink.LinkByName(iface)
+	if err != nil {
+		return err
+	}
+
+	ip, ipnet, err := net.ParseCIDR(ipAddr)
+	if err != nil {
+		return err
+	}
+	ipnet.IP = ip
+
+	if err := netlink.AddrAdd(link, &netlink.Addr{IPNet: ipnet}); err != nil {
+		return err
+	}
+
+	if err := netlink.LinkSetUp(link); err != nil {
+		return err
+	}
+
+	return nil
+}
```
There's very similar code in `pkg/datapath/linux/routing` as well, and probably elsewhere too. If you have the time, it'd be nice to consolidate the implementation in the code. Non-blocking comment tho.

brb replied:
Yeah, it makes sense to consolidate both suites. Once this PR is merged, I will create a good-first-issue for it.

christarazi commented on pkg/node/address.go:
```diff
@@ -44,6 +44,9 @@ var (
 	ipv4AllocRange      *cidr.CIDR
 	ipv6AllocRange      *cidr.CIDR
 
+	// k8s Node IP (either InternalIP or ExternalIP or nil; the former is prefered)
```
Nit: could you update the copyright year? (Github UI doesn't let me select unchanged lines.)

christarazi commented on pkg/option/config.go:
```diff
@@ -2450,6 +2460,26 @@ func (c *DaemonConfig) Populate() {
+			if !found {
+				c.Devices = append(c.Devices, device)
+			}
+		}
```
Nit: non-blocking comment--to keep the code less indented, you could return out of the `if` statement, and convert the `else` block to simply fallthrough

brb committed:
daemon: Disable kube-proxy replacement in unit tests

Otherwise, we would need to make the daemon unit test suite as
privileged due to bpftool depedency for the feature detection in
initKubeProxyReplacementOptions().

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
k8s: Keep waiting for Node object when dev auto-detection might happen

Previously, when running with "--k8s-require-ipv{4,6}-pod-cidr=false",
retrieveNodeInformation() might have failed retrieving a self
(Cilium)Node object on a busy cluster.

To avoid this, return an error if the BPF NodePort dev auto-detection
might happen. This will make the retry mechanism not to give up on
waiting for the object.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb commented on daemon/cmd/daemon.go:
```diff
@@ -413,6 +416,52 @@ func NewDaemon(ctx context.Context, dp datapath.Datapath) (*Daemon, *endpointRes
+		}
+	}
+	if option.Config.EnableHostFirewall && len(option.Config.Devices) == 0 {
+		device, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
```
I left this to make the host-fw to work when BPF NodePort is disabled.

pchaigno commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -370,8 +370,14 @@ data:
 {{- end }}
+{{- if .Values.global.nodePort.devices }}
+  devices: {{ join " " .Values.global.nodePort.devices | quote }}
+{{- end }}
```
Any reason `devices` is under `global.nodePort.` and not just `global.`? It isn't specific to NodePort as far as I understand.

brb replied:
Hmm, good point. BPF Nodeport doesn't depend on host-fw (and vice versa), so it makes sense to define devices via `global.devices` instead. I will change it.

As a side note, I was thinking about introducing more fine grained device selection in v1.9 for each feature, i.e. `global.nodePort.devices`, `global.hostFW.devices`, `global.masquerade.devices`. But in any case, the selection via `global.devices` makes sense.

brb replied:
Resolved in: e361f6121a31140ba2fa9e77abcab6b852bf7014.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -91,19 +91,6 @@ struct dsr_opt_v6 {
-#else
-	return false;
-#endif
-}
```
Was this added to handle some specific potential bug? If we want to keep it, instead of returning `true`, we could return some static data set per device at load time, e.g., `IS_AFTER_XDP_PROG`. Not sure if it's worth it here.

brb replied:
This was just a sanity check. When the XDP is loaded, the `!backend_local` case should have been handled by bpf_xdp. So, any handling of this case in bpf_host is considered as a bug (except for a case when a relevant svc has been added just after bpf_xdp finished handling a request).

joestringer commented on pkg/k8s/init.go:
```diff
@@ -67,14 +67,21 @@ func waitForNodeInformation(ctx context.Context, nodeName string) *nodeTypes.Nod
+	// self object.
+	mightAutoDetectDevices :=
+		option.Config.KubeProxyReplacement != option.KubeProxyReplacementDisabled &&
+			(len(option.Config.Devices) == 0 || option.Config.DirectRoutingDevice == "")
```
nit: Given that `GetNodeSpec()` also needs this logic it seems appropriate for a separate function. I also wonder if it would be more suitable to live near other devices logic, eg in the `option` package.

joestringer commented on daemon/cmd/kubeproxy_replacement_test.go:
```diff
@@ -0,0 +1,165 @@
+	c.Assert(option.Config.Devices, checker.DeepEquals, []string{"dummy0"})
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy0")
+
+	// 4. dummy1 should be detected too
+	c.Assert(addDefaultRoute("dummy1", "192.168.1.1"), IsNil)
+	c.Assert(detectDevices(true, true), IsNil)
+	sort.Strings(option.Config.Devices)
+	c.Assert(option.Config.Devices, checker.DeepEquals, []string{"dummy0", "dummy1"})
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy0")
+
+	// 5. Enable IPv6, dummy1 should not be detected, as no default route for
+	// ipv6 is found
+	option.Config.EnableIPv6 = true
+	c.Assert(detectDevices(true, true), IsNil)
+	c.Assert(option.Config.Devices, checker.DeepEquals, []string{"dummy0"})
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy0")
```
I don't fully follow case 5 here. I presume the state from case 4 (ie ipv4 default route) should remain, so now that we don't have a ipv6 default route we don't care about the device with the ipv4 default route? IPv4 and IPv6 must both have default routes and both route over the same device?

brb replied:
The test case does not set neither IPv6 addr nor a default IPv6 route. Anyway, due to `option.Config.EnableIPv6`,
`linuxdatapath.NodeDeviceWithDefaultRoute()` will hit the case https://github.com/cilium/cilium/blob/pr/brb/auto-multi-dev-v3/pkg/datapath/linux/route/route_linux.go#L566 which will make the former function to skip the default route device detection.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -419,6 +422,52 @@ func NewDaemon(ctx context.Context, dp datapath.Datapath) (*Daemon, *endpointRes
+			log.Fatalf("BPF masquerade works only in veth mode (--%s=\"%s\"", option.DatapathMode, datapathOption.DatapathModeVeth)
+		}
+		if option.Config.EgressMasqueradeInterfaces != "" {
+			log.Fatalf("BPF masquerade does not allow to specify devices via --%s. Use --%s instead.", option.EgressMasqueradeInterfaces, option.Devices)
```
Is this basically because `--egress-masquerade-interfaces` receives an iptables wildcard parameter which can specify either directly a device or a range of devices with syntax like `eth+` and we don't want to support that? Or more just to avoid conflicting configuration from similar options?

joestringer replied:
Oh, I guess this is already in the tree. Well I'm still curious.

brb replied:
> receives an iptables wildcard parameter which can specify either directly a device or a range of devices with syntax like eth+ and we don't want to support that?

Yes, that's the reason. Once we decouple BPF NodePort from BPF masq (in v1.9), I'm planning to add a support for `--egress-masquerade-interfaces` with BPF masq.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1733,9 +1702,8 @@ func initKubeProxyReplacementOptions() {
-				Info("Using auto-derived device for BPF node port")
-			option.Config.Devices = []string{device}
+			log.WithField(logfields.Interface, option.Config.Devices).
+				Info("Using auto-derived devices for BPF node port")
```
Might be nice to get the equivalent message for `DirectRoutingDevice` here as well.

joestringer reviewed (approved):
Just a few minor nits, LGTM.

brb committed:
test: Temporary disable kubeproxy-free on 4.19

Disable the kube-proxy replacement on the CI 4.19 job until
has been merged https://github.com/cilium/cilium/pull/11915, but
keep bpf_sock to avoid bpf_lxc complexity issues.

Also, get rid of non-working kernel vsn setting which complicates
passing of the KERNEL env var to ginkgo test runner.

Finally, disable ipsec + vxlan test on 4.19, as it is broken with the
contemporary 4.19 setup (TODO to investigate it).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm reviewed (approved):
Changes LGTM does this require changes in the upgrade guide or are they already available?

borkmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -253,9 +253,17 @@ func init() {
+	flags.StringP(option.Device, "d", "",
+		fmt.Sprintf("Device facing cluster/external network for attaching bpf_host. Deprecated in favor of --%s", option.Devices))
+	flags.MarkDeprecated(option.Device, "This option will be removed in v1.10")
 	option.BindEnv(option.Device)
```
Can't we remove this right here and describe in the upgrade notes what a user needs to do? Why wait until 1.10? (I doubt many users deploy with an explicit --device.)

brb replied:
We would break the existing deployments in which `--device` was specified. I thought that we have a policy that we can change a user facing flag in a second release after we have announced it? @aanm might know better.

aanm replied:
correct. We can't simply remove flags and hope for the best :-)

borkmann replied:
> correct. We can't simply remove flags and hope for the best :-)

But we eventually are going to remove them. I don't think anyone will read the daemon help. So it's basically the same issue in the second release from now.

aanm replied:
> > correct. We can't simply remove flags and hope for the best :-)
> 
> But we eventually are going to remove them. I don't think anyone will read the daemon help. So it's basically the same issue in the second release from now.

If the users use a deprecated flag it will show up in the logs, also Martynas will add some notes in the upgrade guide.

borkmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -253,9 +253,17 @@ func init() {
 	option.BindEnv(option.Device)
 
+	flags.StringSlice(option.Devices, []string{}, "List of devices facing cluster/external network for attaching bpf_host")
+	option.BindEnv(option.Devices)
```
I would drop the `bpf_host`. A user doesn't/shouldn't know what bpf_host means here. I'd just say `List of devices facing cluster/external network`.

brb replied:
Maybe I would add `(used for BPF NodePort, host firewall and BPF masquerading)` to make it clear.

borkmann replied:
I would make it all explained in detail in the doc, but it's probably also too much context for here (implementation detail).

borkmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -253,9 +253,17 @@ func init() {
+	option.BindEnv(option.Devices)
+
+	flags.String(option.DirectRoutingDevice, "", "Device name used to connect nodes in direct routing mode (required only by BPF NodePort; derived automatically if empty)")
+	option.BindEnv(option.DirectRoutingDevice)
```
I would briefly state from where (`derived automatically if empty`) this is derived.

borkmann replied:
I presume this has to be part of `--devices` list. Perhaps should be stated.

brb replied:
> I presume this has to be part of --devices list. Perhaps should be stated.

It's derived from a device which as a k8s node IP (the device is used by the direct routing routes). If the device was not specified within `--devices`, then it will be added to the list.

borkmann replied:
Given we would have detection working nicely for it, what k8s use-case do you have in mind where we would still need the explicit DirectRoutingDevice specified? (I just think from a user pov, it's confusing.)

brb replied:
When a user manages routing by itself (e.g. via BGP; `--auto-direct-node-routes=false`). Then they might want to control which interface should be used for direct routing.

borkmann replied:
Fair enough, if there are known users that need this control beyond default detection, then lets add. (I just think that all these knobs to specify devices are just limited since it requires all nodes to be exact homogeneous wrt their device [and device naming] setup in order to work.)

borkmann reviewed (approved):
Modulo mentioned nits.

christarazi commented on pkg/node/types/node.go:
```diff
@@ -223,6 +223,22 @@ func (n *Node) getNodeIP(ipv6 bool) (net.IP, addressing.AddressType) {
 
+// GetK8sNodeIPs returns k8s Node IP (either InternalIP or ExternalIP or nil;
+// the former is prefered).
+func (n *Node) GetK8sNodeIP() net.IP {
```
I was looking at the file recently and noticed that we have two very similar functions: `GetK8sNodeIP` and `GetNodeIP` (immediately below). I tracked it down to this PR. I am just curious; why did we need to add this @brb?

brb replied:
`GetNodeIP()` can return an IP which is neither k8s `InternalIP` nor k8s `ExternalIP`. For the device detection, we want to use k8s IPs, as they are used for communication between nodes. For this, I introduced `GetK8sNodeIP()`.
